mcs/class/System/System.Net.Sockets/Socket.cs

```diff
@@ -1250,37 +1314,52 @@ public bool ConnectAsync (SocketAsyncEventArgs e)
 			if (e.RemoteEndPoint == null)
 				throw new ArgumentNullException ("remoteEP");
 
-			e.curSocket = this;
-			e.Worker.Init (this, e, SocketOperation.Connect);
+			SocketAsyncResult sockares = new SocketAsyncResult (this, s => SocketAsyncEventArgsCallback (ConnectAsyncCallback, s), e, SocketOperation.Connect);
 
-			SocketAsyncResult result = e.Worker.result;
+			e.current_socket = this;
+			e.SetLastOperation (SocketOperationToSocketAsyncOperation (SocketOperation.Connect));
+			e.SocketError = SocketError.Success;
+			e.BytesTransferred = 0;
 
 			try {
 				IPAddress [] addresses;
-				IAsyncResult ares;
+				SocketAsyncResult ares;
 
 				if (!GetCheckedIPs (e, out addresses)) {
-					result.EndPoint = e.RemoteEndPoint;
-					ares = BeginConnect (e.RemoteEndPoint, SocketAsyncEventArgs.Dispatcher, e);
+					sockares.EndPoint = e.RemoteEndPoint;
+					ares = (SocketAsyncResult) BeginConnect (e.RemoteEndPoint, s => SocketAsyncEventArgsCallback (ConnectAsyncCallback, s), e);
 				} else {
 					DnsEndPoint dep = (e.RemoteEndPoint as DnsEndPoint);
-					result.Addresses = addresses;
-					result.Port = dep.Port;
-					ares = BeginConnect (addresses, dep.Port, SocketAsyncEventArgs.Dispatcher, e);
+					sockares.Addresses = addresses;
+					sockares.Port = dep.Port;
+					ares = (SocketAsyncResult) BeginConnect (addresses, dep.Port, s => SocketAsyncEventArgsCallback (ConnectAsyncCallback, s), e);
```

Comment:
This delegate pair design is problematic as it does causes allocations on every callback.